vllm/model_executor/models/ultravox.py

```diff
@@ -301,7 +302,15 @@ def forward(
 @MULTIMODAL_REGISTRY.register_max_multimodal_tokens(
     "audio", get_ultravox_max_audio_tokens)
 @MULTIMODAL_REGISTRY.register_processor(UltravoxMultiModalProcessor)
-class UltravoxModel(nn.Module, SupportsMultiModal, SupportsPP):
+class UltravoxModel(nn.Module, SupportsMultiModal, SupportsPP, SupportsLoRA):
+    packed_modules_mapping = {}
+    # llama3 supported modules
+    supported_lora_modules = [
+        'up_proj', 'down_proj', 'gate_proj', 'v_proj', 'o_proj', 'k_proj',
+        'q_proj'
+    ]
```

Comment:
@jeejeelee since you have worked on loras extensively in vllm, i want to get your input. Ultravox uses llama for language model and loads vllm `LlamaForCausalLM` which supports LoRA. Given that, what should happen in Ultravox to support LoRA?  

As I was adding Llama 3.1 supported modules here, I found that vLLM wasn't loading all v_proj, k_proj, etc. Is that because LlamaForCausalLM supports packed module for `vkq_proj`? Also I didn't find `packed_modules` in the original LLama 3.1 model. Why does vLLM defines those in LlamaForCausalLM?

Reply:
Also do i need to use `LLMWrapper` when initializing the language model to align the prefix so that vLLM can load Peft loras? Like this here: https://github.com/vllm-project/vllm/pull/7199/files#diff-7b8a4e258637b7c94389c745c449c52137d33cf92957f3e5bcb18a0ee204b21bR807

Reply:
> Given that, what should happen in Ultravox to support LoRA?

Treat it as a independent model - you need to add LoRA-related static variables to it

>Why does vLLM defines those in LlamaForCausalLM?

Since these are static variables in the model interface

> Also do i need to use LLMWrapper

`LLMWrapper`  has been removed, so this probably doesn't need to be added. I'll figure it out later.

Reply:
For mixtral, should i also add its supported lora modules?

Reply:
How does one figure out what are supported lora modules in LlamaForCausalLM because Llama upstream have no packed modules? Just trying to understand how it all works

Reply:
> Since Ultravox can be run with mixtral, should i also add its supported lora modules?

Could you plz provide the model link？

> How does one figure out what are supported lora modules in LlamaForCausalLM because Llama upstream have no packed modules? Just trying to understand how it all works 

Generally, all linear-related modules support LoRA.

Reply:
> Could you plz provide the model link？
Are you asking about the link to Mixtral model? If yes, then @petersalas do you know which Mixtral model ultravox uses?

> Generally, all linear-related modules support LoRA.
I understand that part but the part I'm having difficulty with is LlamaCausal in vLLM have packed modules where upstream llama model in transformer library doesn't. Why is that? Trying to understand that part better

I also ran more tests with using Lora modules as described in LlamaCausal and I can see from the logs it is not loading kqv packed projections

Reply:
Let's take Query, Key, Value as an example: upstream llama use three separate linear layers for implementation, while vLLM fuses them into a single linear layer called `QKVParallelLinear`,  gate_up_proj` follows similar logic

Reply:
The mistral model is here: https://huggingface.co/fixie-ai/ultravox-v0_4_1-mistral-nemo. AFAICT it _should_ just work in vLLM now that #11198 has landed, but I haven't tried after that change (previously we had to hack up the replacement token which was hardcoded to Llama).

Reply:
@jeejeelee As I'm testing this PR, I'm seeing difference in output when I apply LoRA on Llama vs Llama in Ultravox. It is because vLLM isn't able to activate LoRA modules because names don't match up e.g.  `language_model.model.layers.9.self_attn.qkv_proj` in ultravox vs `model.layers.9.self_attn.qkv_proj` in lora adapter. LLMWrapper was there to resolve this issue but since it has been removed in @DarkLight1337 [PR](https://github.com/vllm-project/vllm/pull/10751/files), do you have any recommendation how to achieve this? 

I also looked at linked PR but solution is specific to Qwen2ForCausalLM. Or am i missing something?

Any concerns if I add a warning in vLLM when this happens to improve user experience? Because most of LoRA code works until the point of activation because lora modules matching happens based on module types (i.e. down_proj, k_proj, etc) and not based on fully qualified name

Reply:
Which LoRA model are you currently using?

Reply:
Do you mean [this lora model](https://github.com/vllm-project/vllm/pull/11253/files#diff-d30d5acc83856e95b5ac61c68a331d06d8d02398b411f5fa055c5d41319c4760R158)?

Reply:
Given that the current multimodal model  only supports adding LoRA to the language backbone, we can temporarily add LoRA as shown in #10022 , If that's the case, you only need to modify a small amount of code, similar to #10022

Reply:
Ohh I think i have all that code if you check my pr. But vllm isn't activating the module because fully qualified module names don't match as per my previous comment. Does that make sense?

Reply:
Treat UltraVox as an independent model, and  its fully qualified module names differ from LLaMA's, you cannot use LLaMA's LoRA.

Reply:
Aren't the qualified module names specific to vLLM? AFAICT I have checked peft and I can't find this limitation. I'm writing up the code for peft to verify that. Maybe I'm missing something but it does seem odd to me that I can't apply an existing LLama Lora in Ultravox

Reply:
@jeejeelee you were right. after further troubleshooting, peft don't apply llama lora on ultravox either - it ignores missing modules silently until the recent version which logs warning if lora is missing model modules.  So I ended up creating a transformed lora for ultravox for testing and added code in vllm to log a warning if lora module is missing for a model. 

I have cleaned up the PR and added a test case which should give enough coverage. I'm checking few more things but the PR should be ready shortly